syntax = "proto3";

message Request {
  oneof request {
    InitRequest           init_request            = 1;
    FetchDataRequest      fetch_data_request      = 2;
    FetchLabelsRequest    fetch_labels_request    = 3;
    FetchResponsesRequest fetch_responses_request = 4;
    string                unused                  = 5;
  }
}

message Response {
  oneof response {
    InitResponse            init_response             = 1;
    FetchDataResponse       fetch_data_response       = 2;
    FetchLabelsResponse     fetch_labels_response     = 3;
    FetchResponsesResponse  fetch_responses_response  = 4;
  }
}

message InitRequest {}
message FetchDataRequest {}
message FetchLabelsRequest {}
message FetchResponsesRequest {}

message InitResponse {
  int32 num_samples = 1;
  int32 data_size = 2;

  bool  has_labels = 3;
  int32 num_labels = 4;
  int32 label_size = 5;

  bool  has_responses = 6;
  int32 num_responses = 7;
  int32 response_size = 8;

  repeated int32 data_dims = 9;

  string reader_type = 10;
}

message FetchDataResponse {
  repeated float data = 1;
}

message FetchLabelsResponse {
  repeated bool labels = 1;
}

message FetchResponsesResponse {
  repeated float responses = 1;
}
